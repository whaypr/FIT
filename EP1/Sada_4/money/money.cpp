/**
*
* Variace na zjednodušený problém batohu (závažíčka)
*
*/

#include <iostream>
#include <vector>

using namespace std;

#define INT_MAX 2147480000
using ll = long long;


ll getTenders(vector<ll>& tenders) {
	ll sum = 0;

	while (true) {
		double tmp_value;
		cin >> tmp_value;
		if (tmp_value == -1) return sum;
		
		ll value = tmp_value * 100 + 0.5;

		int number;
		cin >> number;

		while(number--) {
			tenders.push_back(value);
			sum += value;
		}

		char dummy;
		cin >> dummy;
	}
}


void findAmounts(int amounts[], const vector<ll>& tenders, const ll size) {
	for (const auto & t : tenders) {
		for (long i = size; i >= 0; --i) {
			if (i - t < 0 || amounts[i-t] == INT_MAX) continue;
			amounts[i] = min( amounts[i], amounts[i-t] + 1 );
		}
	}
}


int main() {
	while (true) {
		double tmp_amount;
		cin >> tmp_amount;
		if (tmp_amount == -1) break;

		ll amount = tmp_amount * 100 + 0.5;

		// get tenders
		vector<ll> tendersCustomer, tendersShopkeeper;

		ll sumCustomer = getTenders(tendersCustomer);
		ll sumShopkeeper = getTenders(tendersShopkeeper);

		// find amounts
		int * amountsCustomer = new int[sumCustomer+1];
		int * amountsShopkeeper = new int[sumShopkeeper+1];

		amountsCustomer[0] = amountsShopkeeper[0] = 0;
		for (ll i = 1; i <= sumCustomer; ++i)
			amountsCustomer[i] = INT_MAX;
		for (ll i = 1; i <= sumShopkeeper; ++i)
			amountsShopkeeper[i] = INT_MAX;

		findAmounts(amountsCustomer, tendersCustomer, sumCustomer);
		findAmounts(amountsShopkeeper, tendersShopkeeper, sumShopkeeper);

		// find difference
		int minim = INT_MAX;

		for (ll i = amount; i <= min(sumCustomer, sumShopkeeper + amount); ++i)
			if (amountsCustomer[i] == INT_MAX || amountsShopkeeper[i-amount] == INT_MAX)
				continue;
			else
				minim = min( minim, amountsCustomer[i] + amountsShopkeeper[i-amount] );

		// print result
		if (minim < INT_MAX)
			cout << minim << " tenders must be exchanged." << endl;
		else
			cout << "The payment is impossible." << endl;

		delete [] amountsCustomer;
		delete [] amountsShopkeeper;
	}

	return 0;
}
