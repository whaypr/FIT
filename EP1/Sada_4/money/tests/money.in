100.80
500 1x
200 3x
1.00 10x
0.20 2x
-1
500 10x
200 12x
100 8x
0.10 1x
0.20 0x
0.50 100x
20 2x
-1
200
10 19x
-1
200 1x
-1
393.83
0.15 6x
-1
4.92 3x
-1
166.49
0.27 1x
-1
0.63 0x
-1
139.26
7.36 0x
-1
4.29 1x
-1
157.82
0.23 0x
8.02 7x
-1
31.67 2x
0.42 3x
-1
262.29
7.84 3x
3.15 7x
-1
10.91 0x
18.62 0x
-1
491.70
9.25 6x
15.05 4x
-1
0.13 6x
45.82 7x
-1
5.45
3.64 4x
18.08 3x
5.84 6x
0.54 3x
6.76 9x
12.26 8x
7.95 6x
0.78 0x
29.02 7x
0.52 7x
-1
42.86 6x
34.44 1x
30.31 9x
44.81 7x
0.27 5x
3.53 7x
-1
62.19
0.71 4x
0.19 2x
17.15 0x
7.23 0x
0.46 8x
0.55 1x
32.28 9x
0.93 1x
1.24 0x
7.43 4x
-1
0.65 1x
0.32 9x
32.75 1x
0.21 7x
2.37 8x
44.28 5x
0.28 3x
44.43 9x
36.06 3x
0.18 0x
-1
351.28
9.17 8x
4.70 6x
47.72 3x
25.05 5x
0.86 9x
34.64 9x
38.04 7x
-1
5.68 9x
8.10 0x
0.61 5x
43.20 0x
35.22 6x
32.82 5x
0.37 8x
20.36 2x
0.79 2x
9.73 0x
-1
349.30
1.63 3x
29.96 9x
6.68 9x
0.26 0x
0.90 6x
9.36 4x
1.79 7x
3.48 8x
0.09 2x
25.87 6x
3.72 6x
0.19 1x
9.39 9x
17.05 1x
0.50 8x
42.24 4x
40.81 2x
0.92 0x
6.25 2x
0.99 5x
7.13 0x
-1
43.36 9x
0.94 2x
0.69 4x
0.50 3x
0.42 5x
32.05 9x
6.13 1x
0.26 1x
2.02 9x
28.42 6x
-1
104.98
2.48 6x
6.48 3x
17.46 0x
30.46 8x
19.90 8x
4.97 3x
1.25 3x
1.88 2x
4.60 7x
13.04 4x
7.48 8x
26.97 6x
0.39 9x
0.03 2x
0.47 0x
5.35 8x
2.15 4x
0.29 7x
0.85 9x
1.77 9x
9.49 0x
2.92 9x
0.44 5x
3.40 9x
0.26 1x
0.42 2x
17.61 0x
4.25 1x
-1
0.24 0x
32.05 7x
1.00 8x
0.94 6x
35.29 6x
0.47 6x
6.31 0x
0.27 3x
-1
392.14
4.12 5x
11.89 9x
16.20 4x
3.38 3x
18.56 6x
0.60 7x
0.05 9x
49.45 9x
8.73 3x
16.07 7x
0.14 8x
4.59 1x
3.88 8x
0.57 4x
2.70 1x
5.69 9x
0.94 3x
-1
7.91 2x
11.57 7x
0.47 4x
5.37 1x
40.98 0x
0.16 5x
11.39 6x
45.80 6x
48.62 8x
26.85 2x
-1
348.57
1.27 4x
28.18 0x
0.49 5x
49.34 4x
36.14 1x
49.18 9x
9.82 5x
15.74 7x
2.53 6x
7.13 3x
7.75 9x
0.10 8x
0.17 2x
0.37 6x
25.73 1x
0.91 8x
18.36 0x
-1
6.61 0x
30.73 3x
2.50 4x
40.06 8x
11.20 5x
0.63 7x
0.81 7x
-1
390.36
2.11 5x
0.96 2x
49.84 9x
0.40 2x
0.98 3x
0.47 0x
0.73 0x
0.47 4x
12.32 4x
0.43 4x
2.78 8x
0.62 9x
0.48 3x
1.22 3x
24.57 4x
0.71 5x
0.18 6x
0.53 1x
0.59 3x
0.97 8x
3.34 0x
0.76 8x
0.52 8x
5.06 1x
1.09 0x
0.78 9x
1.14 9x
30.67 8x
-1
0.81 2x
15.90 5x
38.13 4x
0.81 7x
24.33 1x
0.87 2x
4.04 5x
2.97 2x
35.06 2x
5.95 5x
0.97 9x
39.77 3x
0.32 9x
35.63 2x
26.85 0x
9.90 6x
41.16 3x
0.25 0x
24.90 8x
0.18 6x
0.28 3x
9.55 2x
-1
100.08
0.50 2x
41.26 7x
0.51 9x
0.28 9x
6.10 9x
39.76 3x
10.87 9x
0.57 7x
0.17 6x
0.68 1x
17.76 8x
39.54 5x
0.93 8x
0.55 8x
0.45 2x
0.97 0x
3.82 3x
4.39 1x
-1
0.61 7x
0.41 2x
0.50 6x
12.17 2x
0.15 4x
47.90 6x
0.72 3x
38.28 7x
8.56 1x
3.52 0x
5.17 5x
41.12 1x
46.67 0x
0.32 4x
6.79 4x
49.69 1x
8.44 6x
44.65 0x
0.86 4x
5.35 3x
0.93 2x
6.31 0x
5.48 3x
0.13 5x
43.55 3x
1.56 5x
0.23 0x
-1
401.40
0.61 5x
0.50 0x
33.64 0x
39.19 2x
2.38 9x
0.19 1x
19.56 3x
0.66 1x
1.92 1x
0.68 9x
42.07 1x
3.77 8x
9.57 7x
7.97 2x
1.53 6x
3.93 6x
0.30 5x
21.60 9x
42.80 2x
34.80 1x
41.40 6x
0.38 8x
0.31 7x
26.90 0x
-1
26.71 1x
38.99 0x
23.86 4x
7.10 9x
0.33 3x
0.79 3x
23.51 4x
23.30 9x
9.15 6x
6.34 6x
2.87 3x
17.42 3x
0.24 4x
4.18 9x
0.69 3x
17.99 5x
4.35 4x
0.69 7x
18.37 0x
2.05 2x
25.57 0x
8.92 9x
0.05 1x
5.28 3x
0.45 8x
5.51 2x
0.53 9x
-1
185.06
46.64 7x
1.66 1x
4.37 4x
3.12 3x
1.65 1x
0.69 6x
0.22 5x
5.59 1x
5.71 1x
1.72 0x
35.66 1x
1.96 8x
1.86 4x
5.68 9x
26.39 5x
5.90 7x
16.31 9x
-1
12.25 6x
9.72 5x
0.58 6x
0.52 5x
3.39 5x
26.61 1x
23.39 5x
23.30 4x
7.57 7x
32.92 9x
0.79 2x
2.95 0x
9.37 8x
45.84 0x
0.00 1x
4.40 2x
0.35 9x
1.64 4x
0.49 3x
0.39 1x
3.02 4x
0.91 5x
0.84 7x
19.31 1x
0.27 3x
5.81 4x
6.39 3x
0.88 3x
0.80 4x
7.52 8x
0.10 3x
0.66 5x
-1
328.53
45.47 6x
0.86 2x
19.00 2x
0.32 3x
0.51 3x
41.54 7x
0.34 1x
44.28 3x
0.33 7x
10.14 5x
0.00 5x
0.29 7x
0.79 9x
35.85 4x
0.86 5x
6.16 0x
-1
0.95 6x
38.47 3x
10.62 3x
0.63 6x
44.16 0x
0.44 5x
5.18 0x
0.99 5x
0.08 4x
26.86 3x
8.07 2x
35.96 0x
23.24 5x
47.16 2x
1.87 1x
0.81 1x
6.20 0x
6.28 1x
19.42 5x
0.63 0x
5.15 6x
22.90 7x
3.39 0x
2.13 0x
14.75 4x
9.51 3x
1.68 3x
24.99 1x
4.64 7x
-1
233.91
45.87 0x
34.39 7x
33.71 1x
35.34 1x
46.97 3x
3.35 0x
21.32 3x
0.23 9x
26.29 0x
0.22 9x
0.43 0x
0.66 5x
0.53 5x
29.57 0x
0.83 4x
9.97 0x
27.28 5x
27.98 9x
6.60 0x
2.23 3x
20.35 2x
0.08 3x
0.95 2x
3.76 8x
0.03 9x
-1
19.29 7x
22.80 6x
47.47 2x
0.34 4x
0.42 9x
31.71 3x
0.99 3x
0.55 2x
0.71 1x
0.00 3x
0.42 0x
0.04 2x
25.25 7x
46.36 4x
6.14 0x
0.65 3x
4.09 8x
7.86 3x
13.56 4x
6.31 7x
-1
-1
