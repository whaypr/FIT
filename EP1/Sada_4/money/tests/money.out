7 tenders must be exchanged.
The payment is impossible.
The payment is impossible.
The payment is impossible.
The payment is impossible.
The payment is impossible.
The payment is impossible.
The payment is impossible.
5 tenders must be exchanged.
3 tenders must be exchanged.
12 tenders must be exchanged.
11 tenders must be exchanged.
6 tenders must be exchanged.
11 tenders must be exchanged.
10 tenders must be exchanged.
11 tenders must be exchanged.
6 tenders must be exchanged.
12 tenders must be exchanged.
6 tenders must be exchanged.
9 tenders must be exchanged.
8 tenders must be exchanged.
